### ISSUE TITLE
[question] Cross-compiler creation and sysroot

### ISSUE THREAD

**Issue by fouassir** [conan-io/conan#14087]
### What is your question?

Hello, I have a question regarding the creation of a cross-compilation toolchain, and how to make it usable in conan.

Software versions :
* conan 1.59
* Ubuntu Linux 18.04 (python 3.6.9)
* Windows 10 (latest update, python 3.7.9)

What I am trying to achieve is creating two packages :
1. cross-clang, which will contain LLVM with all options set to cross-compile (i.e LLVM_TARGETS_TO_BUILD="AArch64;ARM")
2. myboard-sysroot, which contain the sysroot for a specific board (extracted from the board's SD card/debootstrap)

The first package exports a few variables corresponding to the cross-compilation toolchain :
```python
def package_info(self):
        bin_folder = os.path.join(self.package_folder, "bin")

        self.env_info.CC = os.path.join(bin_folder, "clang")
        self.env_info.CXX = os.path.join(bin_folder, "clang++")
        self.env_info.LD = os.path.join(bin_folder, "lld")
        self.env_info.AR = os.path.join(bin_folder, "llvm-ar")
        self.env_info.AS = os.path.join(bin_folder, "llvm-as")
        self.env_info.NM = os.path.join(bin_folder, "llvm-nm")
        self.env_info.RANLIB = os.path.join(bin_folder, "llvm-ranlib")
        self.env_info.STRIP = os.path.join(bin_folder, "llvm-strip")
```

And so far, it seems it is working correctly.

In the second package, we are trying to create the sysroot, and make CFLAGS, CXXFLAGS and LDFLAGS available :
```python
    def package_info(self):
        self.cpp_info.resdirs=["sysroot"]

        triple = "aarch64-linux-gnu"
        arch = "armv8.2-a"
        cpu = "cortex-a55"

        sysroot = os.path.join(self.package_folder, "sysroot")
        compiler_common = f"--target={triple} -march={arch} -mtune={cpu}"
        compiler_includes = f"-isystem {sysroot}/usr/include/c++/10 -isystem {sysroot}/usr/include/aarch64-linux-gnu/c++/10"

        self.env_info.SYSROOT = sysroot
        self.env_info.CHOST = f"{triple}"
        self.env_info.CFLAGS = f"{compiler_common} --sysroot={sysroot}"
        self.env_info.CXXFLAGS = f"{compiler_common} --sysroot={sysroot} {compiler_includes} -stdlib=libstdc++"
        self.env_info.LDFLAGS = f"{compiler_common} -fuse-ld=lld -stdlib=libstdc++"
```

(The reason why I am trying to export those flags only in the sysroot is that I could add a new board's sysroot and automatically compile for this new board).

In one of my libraries, I use the following profile to cross-compile (-pr:h) :

```ini
[settings]
os=Linux
arch=armv8
compiler=clang
compiler.libcxx=libc++
compiler.version=16
build_type=Release

[options]

[tool_requires]
*: cmake/3.26.4, cross-clang/16.0.4@myuser/stable, myboard-sysroot/1.0.0@myuser/stable, ninja/1.11.1

[conf]
```

Under Ubuntu 18.04, environment variables from both cross-clang and myboard-sysroot are correctly set and generates libraries which runs correctly on our board. However, the same cannot be said under Windows : it fails because CXXFLAGS, CFLAGS and LDFLAGS from myboard-sysroot seems not propagated, and thus features testing of CMake fails.

So my question would be => how to export CFLAGS, CXXFLAGS and LDFLAGS so that cross-compilation works both under Linux and Windows ? I tried using buildinfo_env, but it has the same issue.

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @fouassir 

Thanks for your question.

It seems you are using the legacy ``env_info``, that has already been superseded by ``buildenv_info`` in Conan 1.X (and it has been removed in Conan 2.0). 
You said you tried with ``buildenv_info`` but it would be great to have the issue reported using it. One of the main reasons is that ``buildenv_info`` can define "append" operations like ``self.buildenv_info.append("CXXFLAGS", ....)``.

But regarding Windows, the reasons could be others. It depends on what is your build system there. Because typical CMake projects in windows do NOT listen to environment variables like CXXFLAGS. Conan has other mechanisms to be able to force the propagation, like ``self.conf_info.append("tools.build:cxx_flags", values)``.

It would be great to have a reproducible example, the whole (simplified and minimal) ``myboard-sysroot/1.0.0@myuser/stable`` and the whole (simplified and minimal) "consumer" conanfile.py and CMakeLists.txt and main.cpp or whatever. In that way we can really understand what is happening and what would be the issue.

**Comment by fouassir** [comment]
Hi @memsharded ,

Thanks for the answer, it clears up things a bit. I'll update my recipe to use **self.buildenv_info** and make sure it works on Linux.

Regarding Windows, I am using CMake and Ninja as the generator to do the cross-compilation; I'll try to send a complete example to reproduce the issue, however I won't be able to share with you the sysroot (as it is copied manually from a real board).

**Comment by fouassir** [comment]
Late reply, but we finally managed to make the cross-compilation toolchain work on our side.

This is the conanfile we ended up with :

```
import os
from conan import ConanFile
from conan.tools.cmake import CMake, CMakeToolchain, cmake_layout
from conan.tools.files import download, unzip, rm

required_conan_version = ">=1.59.0"

class LLVMConan(ConanFile):

    name = "llvm"
    version = "16.0.6"
    settings = "os", "arch", "build_type"
    no_copy_source = True

    def layout(self):
        cmake_layout(self)

    def generate(self):

        if self.settings.os == "Windows":
            tc = CMakeToolchain(self, generator="Visual Studio 17 2022")
        else:
            tc = CMakeToolchain(self, generator="Unix Makefiles")

        tc.variables["LLVM_BUILD_TYPE"] = self.settings.build_type

        # Support for cross-compilation only
        tc.variables["LLVM_TARGETS_TO_BUILD"] = "X86;AArch64;ARM"

        # Only clang + linker
        tc.variables["LLVM_ENABLE_PROJECTS"] = "clang;lld"

        tc.variables["LLVM_ENABLE_TERMINFO"] = False

        # Still enable some error checking
        tc.variables["LLVM_ENABLE_ASSERTIONS"] = True

        # EH and RTTI only needed for Python support
        tc.variables["LLVM_ENABLE_EH"] = False
        tc.variables["LLVM_ENABLE_RTTI"] = False

        # Disable 32 bit build
        tc.variables["LLVM_BUILD_32_BITS"] = False

        # Disable terminfo support (avoid some issues)
        tc.variables["LLVM_ENABLE_TERMINFO"] = False

        # defaults to aarch64 target
        tc.variables["LLVM_DEFAULT_TARGET_TRIPLE"] = "aarch64-linux-gnu"
        tc.generate()

    def source(self):
        url = "https://github.com/llvm/llvm-project/archive/refs/tags/llvmorg-16.0.6.tar.gz"
        filename = "llvmorg-16.0.6.tar.gz"

        download( self,
                  url,
                  filename,
                  verify=False,
                  sha1="1e7860f812dd63823306ec60a913d4217a87bcc0",
                  retry = 3,
                  retry_wait = 1 )

        unzip(self, filename, strip_root=True)

        # Do not package downloaded archive
        rm(self, filename, os.getcwd(), recursive=False)

    def build(self):
        cmake = CMake(self)
        cmake.verbose = True
        cmake.configure( build_script_folder=os.path.join(self.source_folder, "llvm") )
        cmake.build()
        cmake.install()

    def build_requirements(self):
        self.build_requires("cmake/3.26.4")

    def package_info(self):
        bin_folder = os.path.join(self.package_folder, "bin")

        self.env_info.PATH.append(bin_folder)

        self.buildenv_info.define_path("CC", os.path.join(bin_folder, "clang"))
        self.buildenv_info.define_path("CXX", os.path.join(bin_folder, "clang++"))
        self.buildenv_info.define_path("LD", os.path.join(bin_folder, "lld-link"))
        self.buildenv_info.define_path("AR", os.path.join(bin_folder, "llvm-ar"))
        self.buildenv_info.define_path("AS", os.path.join(bin_folder, "llvm-as"))
        self.buildenv_info.define_path("NM", os.path.join(bin_folder, "llvm-nm"))
        self.buildenv_info.define_path("RANLIB", os.path.join(bin_folder, "llvm-ranlib"))
        self.buildenv_info.define_path("STRIP", os.path.join(bin_folder, "llvm-strip"))

        # Conan 1
        self.env_info.CC = os.path.join(bin_folder, "clang")
        self.env_info.CXX = os.path.join(bin_folder, "clang++")
        self.env_info.LD = os.path.join(bin_folder, "lld-link")
        self.env_info.AR = os.path.join(bin_folder, "llvm-ar")
        self.env_info.AS = os.path.join(bin_folder, "llvm-as")
        self.env_info.NM = os.path.join(bin_folder, "llvm-nm")
        self.env_info.RANLIB = os.path.join(bin_folder, "llvm-ranlib")
        self.env_info.STRIP = os.path.join(bin_folder, "llvm-strip")

        # Automatic usage of LLD instead of gcc system linker
        self.env_info.LDFLAGS = " -fuse-ld=lld "
```

We had to force the use of the lld linker, and to make a specific case for the build system, when it is on a make-based system or a windows w/ MSVC.

**Comment by memsharded** [comment]
Thanks very much @fouassir for following up and the feedback about your recipe that works!

As a couple of minor things:

- The ``generator="Unix Makefiles"`` no need to be defined, that is already the default in non Windows
- Hardcoding the generator ``generator="Visual Studio 17 2022"`` is generally not recommended, because it will be hiding the settings. If you define a VS 16 compiler in your settings, the binary for the package will report VS 16, but it has still been built with VS 17. If you drop it, and want to force that version into this specific package, you can add ``llvm/*:compiler.version=XXX`` in your "build" profile, and this package (I assume it is created as tool-require ``--build-require``) will use the version that you specify there. Just a minor thing, if it works for you and you know what you are doing, it is ok, just commenting about a non-hardcoded in recipe alternative.

**Comment by fouassir** [comment]
Hello @memsharded ,

To be honest, creating the LLVM recipe to cross-compile was the easy part; the hardest part has been the creation of the sysroot; as we started from a SD card image, we had to extract libraries and headers from the existing file system, then create a resource-only conanfile to embed all those files and then export the right variables. A bit tedious, with lot of manual work, but we made it.

Concerning generators, we still have an issue, this is why we specified the Visual Studio generator in the recipe. In fact by default, with our profile cmake selects the "MinGW Makefiles" generator under windows.

Our profile :
```
[settings]
os=Windows
arch=x86_64
compiler=Visual Studio
compiler.version=17
build_type=Release

[options]

[build_requires]
*/*: cmake/3.27.7

[env]
```

We use this profile both as **build** and **host** profiles, with the following command :

`conan create conanfile.py --build=missing -pr:b profile -pr:h profile -k`

At the end, on windows we get the following error :

```
CMake Error: CMake was unable to find a build program corresponding to "MinGW Makefiles"
```

It seems like CMake does not detect Visual Studio as a build system. I need to check if the installation of my visual studio is correct.
